### PR TITLE
This adds RPC call to set asset path.

### PR DESCRIPTION
**Pull request by aam**
This is needed when doing hot reload to pick up updated assets.

Currently, asset path is set to point to local device copy of assets on restart only. This RPC allows to point to local copy of assets on hot reload too.

### CODE REVIEW

File: shell/common/platform_view_service_protocol.cc [hunk 1]
```diff
@@ -406,4 +408,58 @@ bool PlatformViewServiceProtocol::FlushUIThreadTasks(const char* method,
...
+    // The view existed and the isolate was created. Success.
+    std::stringstream response;
+    response << "{\"type\":\"Success\","
+             << "\"view\":";
+    AppendFlutterView(&response, view_id_as_num, main_port, isolate_name);
+    response << "}";
+    *json_object = strdup(response.str().c_str());
+    return true;
+  }
+  return true;
```
Comment (a-siva) [inline shell/common/platform_view_service_protocol.cc]: isn't this line unreachable? maybe restructure this code as
  if (!view_existed) {
    return ..
  }
  // The view existed and the isolate was created. Success.
  ....
  ....
  return true;
}

File: shell/common/platform_view_service_protocol.cc [hunk 2]
```diff
@@ -406,4 +408,58 @@ bool PlatformViewServiceProtocol::FlushUIThreadTasks(const char* method,
...
+
+bool PlatformViewServiceProtocol::SetAssetBundlePath(const char* method,
+                                            const char** param_keys,
+                                            const char** param_values,
+                                            intptr_t num_params,
+                                            void* user_data,
+                                            const char** json_object) {
+  const char* view_id =
+      ValueForKey(param_keys, param_values, num_params, "viewId");
+  const char* asset_directory =
```
[the remark this answers is not in the document]
Comment (aam) [reply shell/common/platform_view_service_protocol.cc]: Done

File: shell/common/platform_view_service_protocol.cc (hunk 1 above)
Comment (aam) [reply shell/common/platform_view_service_protocol.cc]: Done here and above.

**Comment by aam** [comment]
@a-siva wrote
> Do we need a version of SetAssetBundlePathOnUI for Fuchsia platform also?

I realized I missed mac and ios implementation of PlatformView class, which should be fixed in 1fcad57 commit, but where is Fuchsia implementation that you mentioned?

**Comment by Hixie** [comment]
Can we #ifdef this code out of release builds?

cc @cbracken for review

File: shell/common/engine.cc [hunk 3]
```diff
@@ -279,6 +279,11 @@ void Engine::RunFromSource(const std::string& main,
   RunBundleAndSource(bundle_path, main, packages);
 }
 
+void Engine::SetAssetBundlePath(const std::string& bundle_path) {
+  TRACE_EVENT0("flutter", "Engine::SetAssetBundlePath");
```
Comment (cbracken) [inline shell/common/engine.cc]: Should this be `TRACE_EVENT_INSTANT0`?

File: shell/common/platform_view_service_protocol.cc [hunk 4]
```diff
@@ -406,4 +407,56 @@ bool PlatformViewServiceProtocol::FlushUIThreadTasks(const char* method,
...
+
+bool PlatformViewServiceProtocol::SetAssetBundlePath(const char* method,
+                                                     const char** param_keys,
+                                                     const char** param_values,
+                                                     intptr_t num_params,
+                                                     void* user_data,
+                                                     const char** json_object) {
+  const char* view_id =
+      ValueForKey(param_keys, param_values, num_params, "viewId");
+  if (view_id == NULL) {
```
Comment (cbracken) [inline shell/common/platform_view_service_protocol.cc]: Here and elsewhere: prefer `nullptr`

**Comment by cbracken** [comment]
Running out for the long-weekend, but will pick up the review by Monday at the latest.

File: shell/common/engine.cc (hunk 3 above)
Comment (aam) [reply shell/common/engine.cc]: TRACE_EVENT0 is what is used in other methods here. Should this be different?

**Comment by aam** [comment]
@Hixie wrote:
> Can we #ifdef this code out of release builds?

#ifdef makes code rather messy, I would rather not use them unless I have to. What is the main concern with having this code in the release build? Code size or functionality?

**Comment by Hixie** [comment]
> What is the main concern with having this code in the release build? Code size or functionality?

I mainly worry about security implications. Any code we have in release builds is code that could be used by an attacker. Code intended to help developers is particularly sensitive because code intended to help developers tends to allow invariants to be broken (by design, e.g. "reload the app with this new code", "load these new assets").

(Code size is also an issue, in that our release builds are an order of magnitude or two bigger than we want them as it is, so any byte we can save here or there is one less byte we have to worry about later.)

**Comment by Hixie** [comment]
@cbracken I believe this is waiting for your review.

File: shell/common/engine.cc (hunk 3 above)
Comment (cbracken) [reply shell/common/engine.cc]: Caveat - replying directly; I haven't reviewed the rest of this file to check every case.

`TRACE_EVENT0` emits two events: a begin immediately followed by an end.
`TRACE_EVENT_INSTANT0` just logs an instant event, which is probably what we want in this case.

I'll take a scan through the rest and we can update them if necessary.

Comment (aam) [reply shell/common/engine.cc]: Isn't TRACE_EVENT0 essentially used when we want to capture time it took for particular action to complete? Why wouldn't we want to get that duration for this event?

Comment (cbracken) [reply shell/common/engine.cc]: If you do, then `TRACE_EVENT0` is what you want.

**Comment by cbracken** [comment]
Alright - finally sat down and read through this: LGTM!
lgtm

**Comment by aam** [comment]
@Hixie wrote
> I mainly worry about security implications. Any code we have in release builds is code that could be used by an attacker. Code intended to help developers is particularly sensitive because code intended to help developers tends to allow invariants to be broken (by design, e.g. "reload the app with this new code", "load these new assets").

I really don't like adding ifdefs to the code as it makes code difficult to read and maintain(therefore easier to introduce errors), but I understand the concern, so let me create an issue to track this.

**Pull request merged**